src/uct/ib/dc/dc_mlx5_ep.h

```diff
@@ -298,26 +300,91 @@ uct_dc_mlx5_ep_from_dci(uct_dc_mlx5_iface_t *iface, uint8_t dci_index)
     /* Can be used with dcs* policies only, with rand policy every dci may
      * be used by many eps */
     ucs_assert(!uct_dc_mlx5_iface_is_dci_shared(iface));
-    return iface->tx.dcis[dci_index].ep;
+    return uct_dc_mlx5_iface_dci(iface, dci_index)->ep;
 }
 
 void uct_dc_mlx5_ep_handle_failure(uct_dc_mlx5_ep_t *ep,
                                    struct mlx5_cqe64 *cqe,
                                    ucs_status_t status);
 
+static UCS_F_ALWAYS_INLINE void
+uct_dc_mlx5_init_dci_config(uct_dc_mlx5_dci_config_t *dci_config,
+                            uint8_t path_index, uint8_t max_rd_atomic)
+{
+    dci_config->path_index    = path_index % UCT_DC_MLX5_IFACE_MAX_DCI_POOLS;
+    dci_config->max_rd_atomic = max_rd_atomic;
+}
+
+ucs_status_t static UCS_F_ALWAYS_INLINE
+uct_dc_mlx5_dci_pool_add_dci(uct_dc_mlx5_iface_t *iface, uint8_t pool_index,
+                             uint8_t dci_index)
+{
+    uct_dc_mlx5_dci_pool_t *pool = &iface->tx.dci_pool[pool_index];
+    uct_dc_dci_t *dci            = uct_dc_mlx5_iface_dci(iface, dci_index);
+    ucs_status_t status;
+
+    ucs_assertv(ucs_array_length(&pool->stack) < iface->tx.ndci,
+                "stack length exceeded ndci");
+
+    status = uct_dc_mlx5_iface_create_dci(iface, dci_index, dci);
+    if (status != UCS_OK) {
+        ucs_error("iface %p: failed to create dci %u at pool %u", iface,
+                  dci_index, pool_index);
+        return status;
+    }
+
+    dci->path_index = pool->config.path_index;
+    dci->pool_index = pool_index;
+
+    if (!uct_dc_mlx5_iface_is_dci_rand(iface)) {
+        *ucs_array_append(&pool->stack, return UCS_ERR_NO_MEMORY) = dci_index;
+    }
```

Comment:
IMO better move it outside this function, to uct_dc_mlx5_dci_pool_append_dci
and:
- rename uct_dc_mlx5_dci_pool_add_dci to uct_dc_mlx5_dci_pool_init_dci
- rename unite uct_dc_mlx5_dci_pool_append_dci into uct_dc_mlx5_iface_dci_can_alloc_or_create since it's called only from there